### PR TITLE
En 5804 enable trie pruning on metachain

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**

### CODE REVIEW

File: data/trie/factory/trieCreator.go [hunk 1]
```diff
@@ -94,9 +93,15 @@ func (tc *trieCreator) Create(trieStorageCfg config.StorageConfig, pruningEnable
...
+	snapshotDbCfg := &config.DBConfig{
+		FilePath:          filepath.Join(trieStoragePath, tc.snapshotDbCfg.FilePath),
+		Type:              tc.snapshotDbCfg.Type,
+		BatchDelaySeconds: tc.snapshotDbCfg.BatchDelaySeconds,
+		MaxBatchSize:      tc.snapshotDbCfg.MaxBatchSize,
+		MaxOpenFiles:      tc.snapshotDbCfg.MaxOpenFiles,
+	}
 
-	trieStorage, err := trie.NewTrieStorageManager(accountsTrieStorage, &tc.snapshotDbCfg, ewl)
+	trieStorage, err := trie.NewTrieStorageManager(accountsTrieStorage, snapshotDbCfg, ewl)
```
Comment (iulianpascalau) [inline data/trie/factory/trieCreator.go]: Might pass configs by value

File: data/trie/trieStorageManager.go [hunk 2]
```diff
@@ -293,6 +288,12 @@ func (tsm *trieStorageManager) snapshot(msh marshal.Marshalizer, hsh hashing.Has
 			continue
 		}
 		db := tsm.getSnapshotDb(snapshot.newDb)
+		if check.IfNil(db) {
+			isSnapshotsBufferEmpty, keys = tsm.isSnapshotsBufferEmpty()
+
+			log.Error("trie storage manager: getSnapshotDb error")
```
Comment (iulianpascalau) [inline data/trie/trieStorageManager.go]: why is this an error?

File: process/block/baseProcess_test.go [hunk 3]
```diff
@@ -395,8 +395,8 @@ func TestBaseProcessor_SetAppStatusHandlerOkHandlerShouldWork(t *testing.T) {
 	assert.Nil(t, err)
 }
 
-//------- RevertStateToBlock
-func TestBaseProcessor_RevertStateToBlockRecreateTrieFailsShouldErr(t *testing.T) {
+//------- RevertStateT
```
Comment (iulianpascalau) [inline process/block/baseProcess_test.go]: trimmed comment

File: process/block/metablock.go [hunk 4]
```diff
@@ -969,6 +970,8 @@ func (mp *metaProcessor) CommitBlock(
 	)
 
 	lastMetaBlock := chainHandler.GetCurrentBlockHeader()
+	mp.updateAccountsStateStorage(lastMetaBlock, mp.stateCheckpointModulus)
```
Comment (iulianpascalau) [inline process/block/metablock.go]: Maybe create a function for these 2 instructions as to keep them organized

File: process/block/metablock.go [hunk 5]
```diff
@@ -1107,6 +1110,34 @@ func (mp *metaProcessor) RevertStateToBlock(header data.HeaderHandler) error {
 	return nil
 }
 
+// RevertState recreates the state tries to the root hashes indicated by the provided header
+func (mp *metaProcessor) RevertState(currHeader data.HeaderHandler, prevHeader data.HeaderHandler) error {
+	err := mp.accounts.RecreateTrie(prevHeader.GetRootHash())
```
Comment (iulianpascalau) [inline process/block/metablock.go]: Duplicated code. Why do we need this?

File: process/block/metablock.go [hunk 6]
```diff
@@ -1107,6 +1110,34 @@ func (mp *metaProcessor) RevertStateToBlock(header data.HeaderHandler) error {
...
+	if err != nil {
+		log.Debug("revert peer state with error for header",
+			"nonce", prevHeader.GetNonce(),
+			"validators root hash", prevHeader.GetValidatorStatsRootHash(),
+		)
+
+		return err
+	}
+
+	mp.pruneStateAccountsOnRollback(currHeader, prevHeader)
```
Comment (iulianpascalau) [inline process/block/metablock.go]: Might have extracted this in a separate function. In this way we will have only one function, not 2: RecreateStateTries and RevertState.

File: process/block/shardblock.go [hunk 7]
```diff
@@ -347,6 +347,23 @@ func (sp *shardProcessor) RevertStateToBlock(header data.HeaderHandler) error {
 	return nil
 }
 
+// RevertState recreates the state tries to the root hashes indicated by the provided header
+func (sp *shardProcessor) RevertState(currHeader data.HeaderHandler, prevHeader data.HeaderHandler) error {
+	err := sp.accounts.RecreateTrie(prevHeader.GetRootHash())
```
Comment (iulianpascalau) [inline process/block/shardblock.go]: Duplicate code

File: process/block/baseProcess.go [hunk 8]
```diff
@@ -990,3 +991,149 @@ func getLastSelfNotarizedHeaderByItself(chainHandler data.ChainHandler) (data.He
...
+	}
+
+	errNotCritical = bp.validatorStatisticsProcessor.PruneTrie(prevValidatorStatsRootHash, data.OldRoot)
+	if errNotCritical != nil {
+		log.Debug(errNotCritical.Error())
+	}
+}
+
+func (bp *baseProcessor) pruneStateAccountsOnRollback(currHeader data.HeaderHandler, prevHeader data.HeaderHandler) {
+	if !bp.accounts.IsPruningEnabled() {
```
Comment (iulianpascalau) [inline process/block/baseProcess.go]: duplicate code

File: process/block/baseProcess.go [hunk 9]
```diff
@@ -990,3 +991,149 @@ func getLastSelfNotarizedHeaderByItself(chainHandler data.ChainHandler) (data.He
...
+
+	if _, ok := header.(*block.Header); ok {
+		return process.GetShardHeaderFromStorage(header.GetPrevHash(), bp.marshalizer, bp.store)
+	}
+
+	return nil, process.ErrWrongTypeAssertion
+}
+
+func (bp *baseProcessor) updatePeerStateStorage(finalHeader data.HeaderHandler, checkpointModulus uint) {
+	if check.IfNil(finalHeader) {
```
Comment (iulianpascalau) [inline process/block/baseProcess.go]: duplicate code

File: process/mock/accountsStub.go [hunk 10]
```diff
@@ -3,6 +3,8 @@ package mock
 import (
 	"errors"
 
+	"github.com/ElrondNetwork/elrond-go/data"
+
```
Comment (iulianpascalau) [inline process/mock/accountsStub.go]: remove empty line

File: cmd/node/config/config.toml [hunk 11]
```diff
@@ -291,6 +291,7 @@
    TimeoutMilliseconds = 100
    Version = 0  # Setting 0 means 'use default value'
 
-[StateTrieConfig]
-    RoundsModulus = 100
-    PruningEnabled = true
+[StateTriesConfig]
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply cmd/node/config/config.toml]: Done.

File: data/trie/factory/trieCreator.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply data/trie/factory/trieCreator.go]: Done.

File: data/trie/trieStorageManager.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply data/trie/trieStorageManager.go]: Removed the logging. If db==nil, the error is somewhere else, not here (It is logged in the getSnapshotDb func).

File: process/block/baseProcess_test.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply process/block/baseProcess_test.go]: Done.

File: process/block/metablock.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply process/block/metablock.go]: Done.

File: process/block/metablock.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply process/block/metablock.go]: Refactored.

File: process/block/metablock.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply process/block/metablock.go]: Done.

File: process/block/shardblock.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply process/block/shardblock.go]: Refactored.

File: process/block/baseProcess.go (hunk 8 above)
Comment (BeniaminDrasovean) [reply process/block/baseProcess.go]: Refactored.

File: process/block/baseProcess.go (hunk 9 above)
Comment (BeniaminDrasovean) [reply process/block/baseProcess.go]: Refactored.

File: process/mock/accountsStub.go (hunk 10 above)
Comment (BeniaminDrasovean) [reply process/mock/accountsStub.go]: Done.

File: node/node_test.go [hunk 12]
```diff
@@ -1922,7 +1988,10 @@ func TestStartConsensus_ShardBootstrapperInvalidConsensusType(t *testing.T) {
 		node.WithEpochStartTrigger(&mock.EpochStartTriggerStub{}),
 		node.WithRequestedItemsHandler(&mock.RequestedItemsHandlerStub{}),
 		node.WithBlockProcessor(&mock.BlockProcessorStub{
-			RevertStateToBlockCalled: func(header data.HeaderHandler) error {
+			RevertStateCalled: func(currHeader data.HeaderHandler, prevHeader data.HeaderHandler) error {
+				return nil
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply node/node_test.go]: Done.

File: process/block/baseProcess.go [hunk 13]
```diff
@@ -990,3 +991,149 @@ func getLastSelfNotarizedHeaderByItself(chainHandler data.ChainHandler) (data.He
...
+	}
+
+	prevHeader, errNotCritical := bp.getPreviousHeader(finalHeader)
+	if errNotCritical != nil {
+		log.Debug(errNotCritical.Error())
+		return
+	}
+
+	prevRootHash := prevHeader.GetRootHash()
+	if prevRootHash == nil {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/block/baseProcess.go]: Removed.

File: process/block/baseProcess.go [hunk 14]
```diff
@@ -992,148 +994,91 @@ func getLastSelfNotarizedHeaderByItself(chainHandler data.ChainHandler) (data.He
...
-		return
-	}
-
-	bp.validatorStatisticsProcessor.CancelPrune(prevValidatorStatsRootHash, data.OldRoot)
-
-	errNotCritical := bp.validatorStatisticsProcessor.PruneTrie(validatorStatsRootHash, data.NewRoot)
-	if errNotCritical != nil {
-		log.Debug(errNotCritical.Error())
+func (bp *baseProcessor) getRootHashes(currHeader data.HeaderHandler, prevHeader data.HeaderHandler, identifier state.AccountsDbIdentifier) ([]byte, []byte) {
+	switch identifier {
```
Comment (iulianpascalau) [inline process/block/baseProcess.go]: We might improve this in the future

**Pull request merged**